lib/HLSL/DxilAddPixelHitInstrumentation.cpp

```diff
@@ -0,0 +1,283 @@
+          auto Index = Builder.CreateAdd(XAsInt, YOffset);
+
+          // Step 3: Clamp to size of UAV to prevent TDR if something goes wrong
+          auto CompareToLimit = Builder.CreateICmpUGT(Index, NumPixelsMinusOneArg);
+          ClampedIndex = Builder.CreateSelect(CompareToLimit, NumPixelsMinusOneArg, Index, "Clamped");
+        }
+
+        // Insert the UAV increment instruction:
+        Function* AtomicOpFunc = HlslOP->GetOpFunc(OP::OpCode::AtomicBinOp, Type::getInt32Ty(Ctx));
+        Constant* AtomicBinOpcode = HlslOP->GetU32Const((unsigned)OP::OpCode::AtomicBinOp);
+        Constant* AtomicAdd = HlslOP->GetU32Const((unsigned)DXIL::AtomicBinOpCode::Add);
+        {
+          (void)Builder.CreateCall(AtomicOpFunc, {
+            AtomicBinOpcode,// i32, ; opcode
+            HandleForUAV,   // %dx.types.Handle, ; resource handle
+            AtomicAdd,      // i32, ; binary operation code : EXCHANGE, IADD, AND, OR, XOR, IMIN, IMAX, UMIN, UMAX
+            ClampedIndex,   // i32, ; coordinate c0: index in elements
+            Zero32Arg,      // i32, ; coordinate c1: byte offset into element
+            Zero32Arg,      // i32, ; coordinate c2 (unused)
+            One32Arg        // i32); increment value
+          }, "UAVIncResult");
+        }
+
+        if (AddPixelCost) {
+          // ------------------------------------------------------------------------------------------------------------
+          // Generate instructions to increment a value corresponding to the current pixel in the second half of the UAV, 
+          // by an amount proportional to the estimated average cost of each pixel in the current draw call.
+          // ------------------------------------------------------------------------------------------------------------
+
+          // Step 1: Retrieve weight value from UAV; it will be placed after the range we're writing to
+          Value * Weight;
+          {
+            Function* LoadWeight = HlslOP->GetOpFunc(OP::OpCode::BufferLoad, Type::getInt32Ty(Ctx));
+            Constant* LoadWeightOpcode = HlslOP->GetU32Const((unsigned)DXIL::OpCode::BufferLoad);
+            Constant* OffsetIntoUAV = HlslOP->GetU32Const(NumPixels * 2);
+            auto WeightStruct = Builder.CreateCall(LoadWeight, {
+              LoadWeightOpcode, // i32 opcode
+              HandleForUAV,     // %dx.types.Handle, ; resource handle
+              OffsetIntoUAV,    // i32 c0: index in elements into UAV
+              Zero32Arg         // i32 c1: byte offset into struct
+            }, "WeightStruct");
+            Weight = Builder.CreateExtractValue(WeightStruct, static_cast<uint64_t>(0LL), "Weight");
+          }
+
+          // Step 2: Update write position ("Index") to second half of the UAV 
+          auto OffsetIndex = Builder.CreateAdd(ClampedIndex, NumPixelsArg);
+
+          // Step 3: Increment UAV value by the weight
+          (void)Builder.CreateCall(AtomicOpFunc,{
```

Comment:
It's just explicitly telling the reader that I don't care about the return value.